### PR TITLE
Refactor native extension

### PR DESCRIPTION
**Pull request by d-unsed**
- Add `Transform` and `TryTransform` traits
- Move Rust functions outside `methods!` macro
- Change naming to match conventions

### CODE REVIEW

File: ext/case_transform/src/lib.rs [hunk 1]
```diff
@@ -2,80 +2,128 @@
...
-
-use ruru::{Class, Object, RString, Hash, Array, Symbol, AnyObject, VM};
+use inflector::cases::{camelcase, classcase, kebabcase, snakecase};
+
+use ruru::{Class, Object, VerifiedObject, RString, Hash, Array, Symbol, AnyObject};
 use ruru::types::ValueType;
+use ruru::result::Error as RuruError;
 
-class!(CaseTransform);
+trait Transform: Object {
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:11]: `Transform` describes an interface for transformation. It can be implemented only to Ruby objects due to `Object` constraint.

Then we make type-specific implementations for
- `AnyObject` (no-op)
- `RString`
- `Symbol`
- `Hash`
- `Array`

by implementing `transorm` methods.

File: ext/case_transform/src/lib.rs [hunk 2]
```diff
@@ -2,80 +2,128 @@
...
+            .map(|item| transform(item, transform_function))
+            .collect::<Array>()
+            .to_any_object()
     }
+}
 
-    fn transformSymbol(value: Symbol, transformMethod: &Fn(AnyObject) -> AnyObject) -> Symbol {
-        let transformed = transformMethod(value);
-        Symbol::new(transformed);
+trait TryTransform: Object {
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:67]: `TryTransform` trait itself is needed only to define `try_transform` method on `AnyObject`. Rust does not allow to change the implementation of types that is not owned by current crate.

File: ext/case_transform/src/lib.rs [hunk 3]
```diff
@@ -2,80 +2,128 @@
...
+    fn try_transform<T>(&self,
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:68]: `try_transform<T>` method:
1. Tries to convert current `AnyObject` to the type `T`.
2. :
   - If the conversion is successful it performs a transformation.
   - otherwise, returns an error from `try_convert_to`.

An example of usage of `try_transform` is

``` rust
object.try_transform::<Hash>(key_transform) // if the object is a hash, try to do a hash transformation
```

The interesting part here is the constraints for `T` type: `where T: VerifiedObject + Transform`.

File: ext/case_transform/src/lib.rs [hunk 4]
```diff
@@ -2,80 +2,128 @@
...
-    ) -> AnyObject {
-        match value.unwrap().ty() {
-            ValueType::Array => transformArray(value, objectTransform).to_any_object(),
-            ValueType::Hash => transformHash(value, objectTransform).to_any_object(),
-            ValueType::Symbol => transformSymbol(value, objectTransform).to_any_object(),
-            ValueType::RString => keyTransform(value).to_any_object(),
-            ValueType::Object => value
-        }
+fn transform(object: AnyObject, key_transform: &Fn(String) -> String) -> AnyObject {
+    let result = object.try_transform::<RString>(key_transform)
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:80]: Try to transform the object as a `RString`
Otherwise try to transform the object as a `Symbol`
...
Otherwise try to transform the object as a `AnyObject` (it is any unkown class, do nothing)

File: ext/case_transform/src/lib.rs [hunk 5]
```diff
@@ -2,80 +2,128 @@
...
+
+    if result.is_err() {
+        unreachable!()
     }
 
-    fn toPascalCase(key: String) -> String { to_class_case(to_snake_case(key.unwrap())) }
-    fn toCamelCase(key: String) -> String { to_camel_case(to_snake_case(key.unwrap())) }
-    fn toDashedCase(key: String) -> String { to_kebab_case(to_snake_case(key.unwrap())) }
-    fn toSnakeCase(key: String) -> String { to_snake_case(key.unwrap()) }
+    result.unwrap()
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:86]: We can safely unwrap here, because `result` has `AnyObject` type

File: ext/case_transform/src/lib.rs [hunk 6]
```diff
@@ -2,80 +2,124 @@
...
+methods! (
+    CaseTransform,
+    _itself,
 
-    fn camel(value: AnyObject) -> AnyObject { transform(value.unwrap().to_any_object(), &camel, &toPascalCase) }
-    fn camelLower(value: AnyObject) -> AnyObject { transform(value.unwrap().to_any_object(), &camelLower, &toCamelCase) }
-    fn dash(value: AnyObject) -> AnyObject { transform(value.unwrap().to_any_object(), &dash, &toDashedCase) }
-    fn underscore(value: AnyObject) -> AnyObject { transform(value.unwrap(), &underscore, &toSnakeCase) }
-    fn unaltered(value: AnyObject) -> AnyObject { value.unwrap().to_any_object() }
+    fn camel(object: AnyObject) -> AnyObject { transform(value.unwrap(), &to_pascal_case) }
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:111]: And we can safely unwrap in this methods, because `object`s are `AnyObject`s

File: ext/case_transform/src/lib.rs [hunk 7]
```diff
@@ -2,80 +2,124 @@
...
+    snakecase::to_snake_case(key)
+}
 
-    fn toPascalCase(key: String) -> String { to_class_case(to_snake_case(key.unwrap())) }
-    fn toCamelCase(key: String) -> String { to_camel_case(to_snake_case(key.unwrap())) }
-    fn toDashedCase(key: String) -> String { to_kebab_case(to_snake_case(key.unwrap())) }
-    fn toSnakeCase(key: String) -> String { to_snake_case(key.unwrap()) }
+class!(CaseTransform);
+
+methods! (
```
Comment (d-unsed) [inline ext/case_transform/src/lib.rs:107]: All the functions which are not used directly in the `CaseTransform` class are moved outside `methods!`

**Comment by d-unsed** [comment]
That's what I meant in d-unseductable/ruru#35 :)

I hope it helps!

```
# Running:

....

Finished in 2.039933s, 1.9608 runs/s, 28.9225 assertions/s.
```

File: ext/case_transform/src/lib.rs [hunk 8]
```diff
@@ -2,80 +2,124 @@
...
-// // underscore: snake_case
-use inflector::cases::snakecase::to_snake_case;
-// // camel_lower: camelCase
-use inflector::cases::camelcase::to_camel_case;
-// // camel: ClassCase (PascalCase)
-use inflector::cases::classcase::to_class_case;
-
-use ruru::{Class, Object, RString, Hash, Array, Symbol, AnyObject, VM};
-use ruru::types::ValueType;
+use inflector::cases::{camelcase, classcase, kebabcase, snakecase};
```
Comment (NullVoxPopuli) [inline ext/case_transform/src/lib.rs:5]: this is handy!

File: ext/case_transform/src/lib.rs [hunk 9]
```diff
@@ -2,80 +2,124 @@
...
 
-class!(CaseTransform);
```
Comment (NullVoxPopuli) [inline ext/case_transform/src/lib.rs:17]: why was this removed? I saw it one one of the example repos, though I don't know what its purpose was

File: ext/case_transform/src/lib.rs [hunk 10]
```diff
@@ -2,80 +2,124 @@
...
+use ruru::{Class, Object, VerifiedObject, RString, Hash, Array, Symbol, AnyObject};
+use ruru::types::ValueType;
+use ruru::result::Error as RuruError;
 
-methods! (
-    CaseTransform,
-    itself,
+trait Transform: Object {
```
Comment (NullVoxPopuli) [inline ext/case_transform/src/lib.rs:11]: I need to read up on traits and impls

for myself:
- https://doc.rust-lang.org/book/traits.html
- https://doc.rust-lang.org/book/generics.html looks like impl's are just implementations of a generic struct?

File: ext/case_transform/src/lib.rs (hunk 9 above)
Comment (d-unsed) [reply ext/case_transform/src/lib.rs:17]: The line was moved to the bottom of the file

File: ext/case_transform/src/lib.rs [hunk 11]
```diff
@@ -2,80 +2,124 @@
...
+
+impl Transform for Array {
+    fn transform(&self, transform_function: &Fn(String) -> String) -> AnyObject {
+        // Temp hack to consume &self for iterator
+        let result = unsafe { self.to_any_object().to::<Array>() };
 
-    fn transformSymbol(value: Symbol, transformMethod: &Fn(AnyObject) -> AnyObject) -> Symbol {
-        let transformed = transformMethod(value);
-        Symbol::new(transformed);
+        result.into_iter()
```
Comment (NullVoxPopuli) [inline ext/case_transform/src/lib.rs:60]: arrays are iterable by default?

edit: not*

File: ext/case_transform/src/lib.rs (hunk 8 above)
Comment (NullVoxPopuli) [reply ext/case_transform/src/lib.rs:5]: is there a way to import the methods into the 'global' namespace of the file? so you don't have to do `classcase::to_class_case`?

File: ext/case_transform/src/lib.rs (hunk 7 above)
Comment (NullVoxPopuli) [reply ext/case_transform/src/lib.rs:107]: this makes sense!

File: ext/case_transform/src/lib.rs (hunk 6 above)
Comment (NullVoxPopuli) [reply ext/case_transform/src/lib.rs:111]: legit

File: ext/case_transform/src/lib.rs (hunk 9 above)
Comment (NullVoxPopuli) [reply ext/case_transform/src/lib.rs:17]: :+1:

File: ext/case_transform/src/lib.rs [hunk 12]
```diff
@@ -2,80 +2,124 @@
...
+
+fn to_snake_case(key: String) -> String {
+    snakecase::to_snake_case(key)
+}
 
-    fn toPascalCase(key: String) -> String { to_class_case(to_snake_case(key.unwrap())) }
-    fn toCamelCase(key: String) -> String { to_camel_case(to_snake_case(key.unwrap())) }
-    fn toDashedCase(key: String) -> String { to_kebab_case(to_snake_case(key.unwrap())) }
-    fn toSnakeCase(key: String) -> String { to_snake_case(key.unwrap()) }
+class!(CaseTransform);
```
Comment (NullVoxPopuli) [inline ext/case_transform/src/lib.rs:105]: here it is!

**Comment by NullVoxPopuli** [comment]
@d-unseductable this helps a ton! thanks so much!

**Comment by NullVoxPopuli** [comment]
I have to wait till I get to my personal laptop to try this out, so I have consistent hardware for my benchmarking :-)

File: ext/case_transform/src/lib.rs (hunk 11 above)
Comment (d-unsed) [reply ext/case_transform/src/lib.rs:60]: They kinda are iterable, but after explicit conversion to iterators :) Same for slices, vectors etc

**Comment by d-unsed** [comment]
You're welcome! Please let me know when you get the results

**Comment by NullVoxPopuli** [comment]
Just for some background here is the comparison between pure ruby, and my last successful compile: https://github.com/rails-api/active_model_serializers/pull/1928#issuecomment-248269973

It actually significantly hurt performance. haha.

**Comment by NullVoxPopuli** [comment]
https://github.com/rails-api/active_model_serializers/pull/1928#issuecomment-248767037

So, slightly faster... but still am not sure what the deal is with the slowness. Still, I appreciate you teaching me rust!

**Comment by NullVoxPopuli** [comment]
I guess what could be next is seeing if fiddle is slowing me down.
